File: frontend_v2/src/types/capabilities.ts
import type { PaginatedPayload } from './api'

export interface CapabilityArticleReference {
  id: number
  title: string
  slug: string
  publish_status: number
  quality_score: number
  updated_at: string | null
}

export interface CapabilitySummaryItem {
  id: number
  group_code: string
  group_name: string
  capability_code: string
  capability_name: string
  category: string | null
  public_claim: string | null
  claim_level: string
  confidence_score: number
  is_active: boolean
  source_count: number
  application_tags: string[]
  updated_at: string | null
}

export interface CapabilityDetailItem extends CapabilitySummaryItem {
  metric_type: string
  unit: string | null
  comparator: string | null
  conservative_value_num: number | null
  conservative_value_text: string | null
  advanced_value_num: number | null
  advanced_value_text: string | null
  internal_note: string | null
  conditions_text: string | null
  recent_articles: CapabilityArticleReference[]
}

export interface CapabilitySourceItem {
  id: number
  source_code: string
  source_vendor: string
  source_title: string
  source_type: string
  source_url: string
  publish_org: string | null
  observed_on: string | null
  reliability_score: number
  citation_note: string | null
  priority_weight: number
}

export interface CapabilityListPayload extends PaginatedPayload<CapabilitySummaryItem> {
  active_total: number
  inactive_total: number
  groups_total: number
}

export interface CapabilityDetailPayload {
  capability: CapabilityDetailItem | null
  warning: string | null
}

export interface CapabilitySourcesPayload {
  spec_id: number
  items: CapabilitySourceItem[]
  warning: string | null
}

export interface CapabilityListFilters {
  active?: boolean
  group_code?: string
  query?: string
  limit?: number
  offset?: number
}

export type CapabilityDisableResult = Record<string, unknown>

File: backend/app/schemas/capabilities.py
"""Capability-related schemas."""

from datetime import datetime

from pydantic import BaseModel, Field


class CapabilityArticleReference(BaseModel):
    """Recent article related to a capability."""

    id: int
    title: str
    slug: str
    publish_status: int
    quality_score: int
    updated_at: datetime | None = None


class CapabilitySummaryItem(BaseModel):
    """Single capability row for list pages."""

    id: int
    group_code: str
    group_name: str
    capability_code: str
    capability_name: str
    category: str | None = None
    public_claim: str | None = None
    claim_level: str
    confidence_score: float
    is_active: bool = True
    source_count: int = 0
    application_tags: list[str] = Field(default_factory=list)
    updated_at: datetime | None = None


class CapabilityDetailItem(CapabilitySummaryItem):
    """Capability detail payload."""

    metric_type: str
    unit: str | None = None
    comparator: str | None = None
    conservative_value_num: float | None = None
    conservative_value_text: str | None = None
    advanced_value_num: float | None = None
    advanced_value_text: str | None = None
    internal_note: str | None = None
    conditions_text: str | None = None
    recent_articles: list[CapabilityArticleReference] = Field(default_factory=list)


class CapabilitySourceItem(BaseModel):
    """Source linked to a capability."""

    id: int
    source_code: str
    source_vendor: str
    source_title: str
    source_type: str
    source_url: str
    publish_org: str | None = None
    observed_on: str | None = None
    reliability_score: float
    citation_note: str | None = None
    priority_weight: int


class CapabilityListPayload(BaseModel):
    """Paginated capability list payload."""

    items: list[CapabilitySummaryItem] = Field(default_factory=list)
    total: int
    limit: int
    offset: int
    active_total: int = 0
    inactive_total: int = 0
    groups_total: int = 0
    warning: str | None = None


class CapabilityDetailPayload(BaseModel):
    """Single capability response."""

    capability: CapabilityDetailItem | None = None
    warning: str | None = None


class CapabilitySourcesPayload(BaseModel):
    """Capability sources response."""

    spec_id: int
    items: list[CapabilitySourceItem] = Field(default_factory=list)
    warning: str | None = None
